File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-12.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        stack = []
        res = 0
        for i in range(len(heights)):
            while stack and heights[i] <= heights[stack[-1]]:
                h = heights[stack.pop()]
                w = i if not stack else i - stack[-1] - 1
                res = max(res, h * w) 
            stack.append(i)

        
        while stack:
            h = heights[stack.pop()]
            w = len(heights) if not stack else i - stack[-1]
            res = max(res, h * w) 
        return res

File: Data Structures & Algorithms/regular-expression-matching/submission-12.cpp
class Solution {
public:
    bool isMatch(string s, string p) {
        int n = s.size(), m = p.size();
        auto back = [&](auto self, int i, int j){
            if (i == n && j == m) {
                return true;
            }
            else if (i > n || j > m) return false;
            if (s[i] == p[j] || p[j] == '.') {
                return self(self, i + 1, j + 1);
            }
            if (p[j] == '*') {
                bool res{false};
                for (int l{}; l <= n - m + j; ++l) {
                    res = res || self(self, i + l, j + 1);
                }
                return res;
            }
            return false;
        };

        return back(back, 0, 0);
    }
};
